src/live-programming.rst

```diff
@@ -20,9 +20,331 @@ Data Analysis Environments
+in machine or program state (e.g., a wrong value in a CPU register,
+an uninitialized object) that lead to the noncompliant behavior.
+Finally, causing the runtime faults are the **software errors** in
+the source code.
+
+Using these terms, debugging may be understood
+more precisely as the collective processes of determining what runtime
+faults led to a runtime failure, determining what software errors
+led to those faults, and modifying the code to prevent the faults from
+occurring. A **debugger** is a tool that allows the programmer to observe
+the machine state during program execution, thereby making it possible to
+identify runtime faults.
+
+Theories of Human Debugging
+---------------------------
+
+Debugging is a complex activity. Recent work develops frameworks and
+theories for understanding the cognitive processes underlying
+programmer errors and debugging strategies, and the ways that
+debugging interfaces influence those processes.
+
+.. container:: bib-item
+
+  .. bibliography:: debugging.bib
+    :filter: key == 'ko2005framework'
+
+  Noting the inadequacy of existing HCI frameworks at the time
+  for analyzing error-prone processes like programming, this paper
+  proposed a framework and methodology for analyzing and designing
+  programming systems that focuses specifically on software errors.
+  The framework deconstructs the various cognitive breakdowns that may
+  occur throughout software development and characterizes software
+  errors in terms of the *chains of cognitive breakdowns* that lead to
+  them. Cognitive breakdowns are characterized, in turn, by their breakdown type,
+  the action being performed by the programmer, the interface being
+  used to perform the action, and the information being acted upon.
+  For example, an interruption may cause a programmer to experience
+  an inattention breakdown upon creating a loop header in their program
+  editor and forget to include the closing brace, introducing a syntax
+  error.
+  The paper described the application of this methodology to
+  evaluate the Alice programming environment by coding and quantifying
+  users' common breakdown chains.
+  The use of this framework directly inspired the design of new programming
+  tools such as the Whyline :cite:`ko2008whyline`.
+
+.. container:: bib-item
+
+  .. bibliography:: debugging.bib
+    :filter: key == 'lawrance2013foraging'
+
+  Prior to this paper, theories of debugging relied on complex
+  mental constructs and were
+  mostly developed in an age in which programming environments were
+  relatively simple. These theories did not take into account the substantial
+  navigational activities performed in modern IDEs and offered little
+  practical advice to builders of software engineering tools.
+  This paper presented a theory of programmer navigation when debugging
+  based on information foraging theory.
+  Information foraging theory is an instance of rational analysis, a
```

Comment:
This definition should probably not be in the summary of the lawrence paper -- instead find the paper that they based it on and make it a bib-item